share/wake/lib/core/tree.wake

```diff
@@ -31,431 +32,622 @@ def ratioQ = 3
-    Tip         = Some (Pair z (i-1))
-    Bin s l x r = if f x then someL z i l else someR x (i+s) r
-  none root
-
-# Extract all elements from the tree which are equal to y
-# => Pair (matches: List x) (rank: Integer)
-export def tequal y (Tree cmp root) =
-  def helperR i out = match _ # i = size including self
-    Tip         = Pair out i
-    Bin s l x r = match (cmp x y)
-      LT = helperR i out r
-      GT = helperL (i-s) out l
-      EQ = helperL (i-s) (x, helperR i out r | getPairFirst) l
-  def helperL i out = match _ # i = size left of self
-    Tip         = Pair out i
-    Bin s l x r = match (cmp x y)
-      LT = helperR (i+s) out r
-      GT = helperL i out l
-      EQ = helperL i (x, helperR (i+s) out r | getPairFirst) l
-  helperL 0 Nil root
+    def upper_ = match _
+        Tip =
+            None
+        Bin _ l x r =
+            def lesserResult =
+                if f x then
+                    Some (Pair x (size l))
+                else
+                    upper_ l
+            upper_ r
+            | omap (editPairSecond (size l + 1 + _))
+            | orElse lesserResult
+    upper_ root
+
+# Extract all elements from the tree which are equal to y, along with their highest rank.
+export def tequal (y: a) (Tree cmp root: Tree a): Pair (List a) Integer=
+    def helperR i out = match _ # i = size including self
+        Tip =
+            Pair out i
+        Bin s l x r = match (cmp x y)
+            LT =
+                helperR i out r
+            GT =
+                helperL (i-s) out l
+            EQ =
+                helperL (i-s) (x, helperR i out r | getPairFirst) l
+    def helperL i out = match _ # i = size left of self
+        Tip =
+            Pair out i
+        Bin s l x r = match (cmp x y)
+            LT =
+                helperR (i+s) out r
+            GT =
+                helperL i out l
+            EQ =
+                helperL i (x, helperR (i+s) out r | getPairFirst) l
+    helperL 0 Nil root
 
 # Returns True if x is an element of y, False otherwise.
-export def x ∈ y = tcontains x y
+export def (x: a) ∈ (y: Tree a): Boolean =
+    tcontains x y
 
 # Returns True if x is NOT an element of y, False otherwise.
-export def x ∉ y = ! x ∈ y
+export def (x: a) ∉ (y: Tree a): Boolean =
+    ! x ∈ y
 
 # Returns True if x contains y, False otherwise.
-export def x ∋ y = y ∈ x
+export def (x: Tree a) ∋ (y: a): Boolean =
+    y ∈ x
 
 # Returns True if x does NOT contain y, False otherwise.
-export def x ∌ y = y ∉ x
+export def (x: Tree a) ∌ (y: a): Boolean =
+    y ∉ x
 
 export def tcontains (y: a) (t: Tree a): Boolean =
-  match t (tupperLE y t)
-    (Tree _   _) None              = False
-    (Tree cmp _) (Some (Pair x _)) = match (cmp x y)
-      EQ = True
-      _  = False
+    match t (tupperLE y t)
+        _ None =
+            False
+        (Tree cmp _) (Some (Pair x _)) = match (cmp x y)
+            EQ =
+                True
+            _ =
+                False
```

Comment:
To describe some of where I'm coming from, I've grown to quite dislike vertical alignment by spaces almost purely due to the fact that it ties mostly-unrelated lines together.  As soon as we decide to rename `x` to be more descriptive, or if the return stops being a `Some (Pair _ _)`, we have to change the otherwise-fine `None` line to match the new length.  Git isn't quite as sensitive to that as some other VCS, but it still adds a lot of unnecessary noise to diffs and blames.